File: Documentation/5.4/Samples/nodejs/ClientApi/Session/Querying/sortQueryResults.js
import { DocumentStore } from "ravendb";

const documentStore = new DocumentStore();
const session = documentStore.openSession();

async function sortQueryResults() {
    {
        //region sort_1
        const products = await session
             // Make a dynamic query on the 'products' collection    
            .query({ collection: "products" })
             // Apply filtering (optional)
            .whereGreaterThan("UnitsInStock", 10)
             // Call 'orderBy'
             // Pass the document-field by which to order the results and the ordering type
            .orderBy("UnitsInStock", "Long")
            .all(); 

        // Results will be sorted by the 'UnitsInStock' value in ascending order,
        // with smaller values listed first.
        //endregion
    }

    {
        //region sort_2
        const products = await session
            .query({ collection: "products" })
             // Apply filtering
            .whereLessThan("UnitsInStock", 5)
            .orElse()
            .whereEquals("Discontinued", true)
             // Call 'orderByScore'
            .orderByScore()
            .all();

        // Results will be sorted by the score value
        // with best matching documents (higher score values) listed first.
        //endregion
    }

    {
        //region sort_3
        const products = await session
            .query({ collection: "products" })
            .whereGreaterThan("UnitsInStock", 10)
             // Call 'randomOrdering'
            .randomOrdering()
             // An optional seed can be passed, e.g.:
             // .randomOrdering("someSeed")
            .all();

        // Results will be randomly ordered.
        //endregion
    }

    {
        //region sort_4
        const numberOfProductsPerCategory = await session
            .query({ collection: "products" })
             // Group by category
            .groupBy("Category")
            .selectKey("Category")
             // Count the number of product documents per category
            .selectCount()
             // Order by the count value
            .orderBy("count", "Long")
            .all();

        // Results will contain the number of Product documents per category
        // ordered by that count in ascending order.
        //endregion
    }

    {
        //region sort_5
        const numberOfUnitsInStockPerCategory = await session
            .query({ collection: "products" })
             // Group by category
            .groupBy("Category")
            .selectKey("Category")
             // Sum the number of units in stock per category
            .selectSum(new GroupByField("UnitsInStock", "sum"))
             // Order by the sum value
            .orderBy("sum", "Long")
            .all();

        // Results will contain the total number of units in stock per category
        // ordered by that number in ascending order.
        //endregion
    }

    {
        //region sort_6
        const products = await session
            .query({ collection: "products" })
             // Call 'OrderBy', order by field 'QuantityPerUnit'
             // Pass a second param, requesting to order the text alphanumerically
            .orderBy("QuantityPerUnit", "AlphaNumeric")
            .all();
        //endregion

        //region sort_6_results
        // Running the above query on the NorthWind sample data,
        // would produce the following order for the QuantityPerUnit field:
        // ================================================================
        
        // "1 kg pkg."
        // "1k pkg."
        // "2 kg box."
        // "4 - 450 g glasses"
        // "5 kg pkg."
        // ...
        
        // While running with the default Lexicographical ordering would have produced:
        // ============================================================================
        
        // "1 kg pkg."
        // "10 - 200 g glasses"
        // "10 - 4 oz boxes"
        // "10 - 500 g pkgs."
        // "10 - 500 g pkgs."
        // ...
        //endregion
    }

    {
        //region sort_7
        const products = await session
            .query({ collection: "products" })
            .whereGreaterThan("UnitsInStock", 10)
             // Apply the primary sort by 'UnitsInStock'
            .orderByDescending("UnitsInStock", "Long")
             // Apply a secondary sort by the score
            .orderByScore()
             // Apply another sort by 'Name'
            .orderBy("Name")
            .all();

        // Results will be sorted by the 'UnitsInStock' value (descending),
        // then by score,
        // and then by 'Name' (ascending).
        //endregion
    }

    {
        //region sort_8
        const products = await session
            .query({ collection: "products" })
            .whereGreaterThan("UnitsInStock", 10)
             // Order by field 'UnitsInStock', pass the name of your custom sorter class
            .orderBy("UnitsInStock", { sorterName: "MySorter" })
            .all();

        // Results will be sorted by the 'UnitsInStock' value
        // according to the logic from 'MySorter' class
        //endregion
    }

    {
        //region get_score_from_metadata
        // Make a query:
        // =============

        const employees = await session
            .query({ collection: "Employees"})
            .search('Notes', 'English')
            .search('Notes', 'Italian')
            .boost(10)
            .all();

        // Get the score:
        // ==============

        // Call 'getMetadataFor', pass an entity from the resulting employees list
        const metadata = session.advanced.getMetadataFor(employees[0]);

        // Score is available in the '@index-score' metadata property
        const score = metadata[CONSTANTS.Documents.Metadata.INDEX_SCORE];
        //endregion
    }
}

{
    //region syntax
    // orderBy overloads:
    orderBy(field);
    orderBy(field, ordering);
    orderBy(field, options);

    // orderByDescending overloads:
    orderByDescending(field);
    orderByDescending(field, ordering);
    orderByDescending(field, options);
    //endregion
}

File: Documentation/5.4/Samples/python/ClientApi/Session/Querying/SortQueryResults.py
from typing import Union, TypeVar

from ravendb import OrderingType, GroupByField, DocumentQuery, SearchOperator, primitives
from ravendb.infrastructure.orders import Product
from ravendb.primitives import constants

from examples_base import ExampleBase, Employee

_T = TypeVar("_T")


class SortQueryResult(ExampleBase):
    def setUp(self):
        super().setUp()
        with self.embedded_server.get_document_store("SortQueryResults") as store:
            with store.open_session() as session:
                self.add_products(session)
                self.add_employees(session)

    def test_sort_query_results(self):
        with self.embedded_server.get_document_store("SortQueryResults") as store:
            with store.open_session() as session:
                # region sort_1
                products = list(
                    session
                    # Make a dynamic query on the 'Products' collection
                    .query_collection("Products")
                    # Apply filtering (optional)
                    .where_greater_than("units_in_stock", 10)
                    # Call 'order_by'
                    # Pass the document-field by which to order the results and the ordering type
                    .order_by("units_in_stock", OrderingType.LONG)
                )

                # Results will be sorted by the 'units_in_stock' value in ascending order,
                # with smaller values listed first
                # endregion

            with store.open_session() as session:
                # region sort_4
                products = list(
                    session.query_collection("Products")
                    # Apply filtering
                    .where_less_than("units_in_stock", 5)
                    .or_else()
                    .where_equals("discontinued", True)
                    # Call 'order_by_score'
                    .order_by_score()
                )

                # Results will be sorted by the score value
                # with best matching documents (higher score values) listed first.
                # endregion

            with store.open_session() as session:
                # region sort_7
                products = list(
                    session.query_collection("Products").where_greater_than("units_in_stock", 10)
                    # Call 'random_ordering'
                    .random_ordering()
                    # An optional seed can be passed, e.g.:
                    # .random_ordering("someSeed")
                )

                # Results will be randomly ordered
                # endregion

            with store.open_session() as session:
                # region sort_10
                number_of_products_per_category = list(
                    session.query_collection("Products", Product)
                    # Group by category
                    .group_by("category").select_key("category")
                    # Count the number of product documents per category
                    .select_count()
                    # Order by the count value
                    .order_by("count", OrderingType.LONG)
                )

                # Results will contain the number of Product documents per category
                # ordered by that count in ascending order.
                # endregion

            with store.open_session() as session:
                # region sort_13
                number_of_units_in_stock_per_category = list(
                    session.query_collection("Products", Product)
                    # Group by category
                    .group_by("category").select_key("category")
                    # Sum the number of units in stock per category
                    .select_sum(GroupByField("units_in_stock", "sum"))
                    # Order by the sum value
                    .order_by("sum", OrderingType.LONG)
                )

                # Results will contain the total number of units in stock per category
                # ordered by that number in ascending order.
                # endregion

            with store.open_session() as session:
                # region sort_16
                products = list(
                    session.query_collection("products")
                    # Call 'order_by', order by field 'quantity_per_unit'
                    # Pass a second param, requesting to order the text alphanumerically
                    .order_by("quantity_per_unit", OrderingType.ALPHA_NUMERIC)
                )

                # endregion

                # region sort_6_results
                #  Running the above query on the NorthWind sample data,
                #  would produce the following order for the quantity_per_unit field:
                #  ================================================================
                #
                #  "1 kg pkg."
                #  "1k pkg."
                #  "2 kg box."
                #  "4 - 450 g glasses"
                #  "5 kg pkg."
                #  ...
                #
                #  While running with the default Lexicographical ordering would have produced:
                #  ============================================================================
                #
                #  "1 kg pkg."
                #  "10 - 200 g glasses"
                #  "10 - 4 oz boxes"
                #  "10 - 500 g pkgs."
                #  "10 - 500 g pkgs."
                #  ...

                # endregion

            with store.open_session() as session:
                # region sort_19
                products = list(
                    session.query_collection("Products").where_greater_than("units_in_stock", 10)
                    # Apply the primary sort by 'units_in_stock'
                    .order_by_descending("units_in_stock", OrderingType.LONG)
                    # Apply a secondary sort by the score
                    .order_by_score()
                    # Apply another sort by 'Name'
                    .order_by("name")
                )

                #  Results will be sorted by the 'units_in_stock' value (descending),
                #  then by score,
                #  and then by 'name' (ascending).
                # endregion

            with store.open_session() as session:
                try:
                    # region sort_22
                    products = list(
                        session.query(object_type=Product).where_greater_than("units_in_stock", 10)
                        # Order by field 'units_in_stock', pass the name of your custom sorter class
                        .order_by("units_in_stock", "MySorter")
                    )

                    # Results will be sorted by the 'units_in_stock' value
                    # according to the logic from 'MySorter' class
                    # endregion
                except RuntimeError as e:
                    pass

            with store.open_session() as session:
                # region get_score_from_metadata
                # Make a query:
                # =============
                employees = list(
                    session.query(object_type=Employee)
                    .search("last_name", "Doe")
                    .search("first_name", "Jane")
                    .boost(10)
                )

                # Get the score:
                # ==============

                # Call 'get_metadata_for', pass an entity from the resulting employee list
                metadata = session.advanced.get_metadata_for(employees[0])

                # Score is available in the '@index-score' metadata property
                score = metadata[constants.Documents.Metadata.INDEX_SCORE]
                # endregion

            class LmaoAPI:
                # region syntax
                # order_by:
                def order_by(
                    self, field: str, sorter_name_or_ordering_type: Union[str, OrderingType] = OrderingType.STRING
                ) -> DocumentQuery[_T]: ...

                # order_by_descending:
                def order_by_descending(
                    self, field: str, sorter_name_or_ordering_type: Union[str, OrderingType] = OrderingType.STRING
                ) -> DocumentQuery[_T]: ...

                # endregion
